contracts/sdk/IpcContract.sol

```diff
@@ -0,0 +1,112 @@
+
+    // List of messages in-flight for which the contract hasn't received a receipt yet.
+    mapping(bytes32 => IpcEnvelope) public inflightMsgs;
+
+    constructor(address gatewayAddr_) {
+        gatewayAddr = gatewayAddr_;
+    }
+
+    /// @notice Entrypoint for IPC-enabled contracts. This function is always called by
+    /// the gateway when a `Call` or `Receipt` cross-net messages is targeted to
+    /// a specific address in the subnet.
+    function handleIpcMessage(IpcEnvelope calldata envelope) external payable onlyGateway returns (bytes memory) {
+        // internal dispatch of the cross-net message to the right method.
+        if (envelope.kind == IpcMsgKind.Call) {
+            CallMsg memory call = abi.decode(envelope.message, (CallMsg));
+            return _handleIpcCall(envelope, call);
+        } else if (envelope.kind == IpcMsgKind.Receipt) {
+            ResultMsg memory result = abi.decode(envelope.message, (ResultMsg));
+
+            // Recover the original message.
+            // If we were not tracking it, or if some details don't match, refuse to handle the receipt.
+            IpcEnvelope storage orig = inflightMsgs[result.id];
+            if (
+                orig.message.length == 0 ||
+                keccak256(abi.encode(envelope.from)) != keccak256(abi.encode(orig.to))
+            ) {
+                revert IpcHandler.UnrecognizedReceipt();
+            }
+
+            /// Note: if the result handler reverts, we will
+            _handleIpcResult(orig, envelope, result);
+            delete inflightMsgs[result.id];
+            return EMPTY_BYTES;
+        }
+        revert UnsupportedMsgKind();
+    }
+
+    /// @notice Function to be overridden by the child contract to handle incoming IPC calls.
+    ///
+    /// NOTE: It's fine for this method to revert. If that happens, IPC will carry the error to the caller.
+    function _handleIpcCall(IpcEnvelope memory envelope, CallMsg memory callMsg) internal virtual returns (bytes memory);
+
+    /// @notice Function to be overridden by the child contract to handle results from previously performed IPC calls.
+    ///
+    /// NOTE: This must not revert as doing so will leave the correlation map in an inconsistent state.
+    /// (IPC will consider the result delivery attempted, and will not repeat it again).
+    function _handleIpcResult(IpcEnvelope storage original, IpcEnvelope memory result, ResultMsg memory resultMsg) internal virtual;
+
+    /// @notice Method the implementation of this contract can invoke to perform an IPC call.
+    function performIpcCall(IPCAddress calldata to, CallMsg calldata callMsg, uint256 value) internal virtual {
+        // Queue the cross-net message for dispatch.
+        IpcEnvelope memory envelope = IGateway(gatewayAddr).sendContractXnetMessage{value: value}(IpcEnvelope({
+            kind: IpcMsgKind.Call,
+            from: to, // TODO: will anyway be replaced by sendContractXnetMessage.
```

Comment:
just a thought, maybe we can create a `callXnetContract` method that hides `kind`, `from` and `nonce`, so that don't need the TODO comment?

Reply:
100% — I captured this as a TODO in #613. I already got started on this but the PR was becoming too heavy. We can address later.

Reply:
(We can also remove the `value`)